[FIXES #6653] Mail notifications for private datasets are public

gannebamm opened the pull request:
closes #6653 

## Checklist
For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [x] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

afabiani reviewed (commented):
@gannebamm I did few changes to the "exclude" query which was not correct IMHO. Could you please give it a try?

gannebamm commented:
This does not work as intended.
On document creation:
```py
Traceback (most recent call last):
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-15-722359ac9ef1>", line 1, in <module>
    user.has_perm('base.view_resourcebase', resource)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/contrib/auth/models.py", line 268, in has_perm
    return _user_has_perm(self, perm, obj)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/contrib/auth/models.py", line 182, in _user_has_perm
    if backend.has_perm(user, perm, obj):
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/guardian/backends.py", line 90, in has_perm
    raise WrongAppError("Passed perm has app label of '%s' while "
guardian.exceptions.WrongAppError: Passed perm has app label of 'base' while given obj has app label 'documents' and given objcontent_type has app label 'documents'
```

and it should be `exclude_users_ids.append(user.id)` (not '[user.id]'). I will work upon this further in the next hours and debugging the hell out of it 😄 .

afabiani commented:
I guess you need to check the polymorphic ctype in order to be sure you are going to exclude the resources correctly.

Let me know if I need to jump in.

gannebamm commented:
I will push my current code to this Branch in a minute. There are still two remaining issues:

**1)** on layer creation, the layer name is not yet set. debug into this:
/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py:421
You will see the name is given in the `**kwargs` but the `obj` ist still titled as `None`. This results in mails like:
```
You have received the following notice from example.com:
The user uploaded the following layer:
You can visit the layer's detail page here: http://example.com/layers/None
To change how you receive notifications, please go to http://example.com 
```

**2)** on Document _upload_ the perms can not be checked. They fail due to `guardian.exceptions.WrongAppError`, or do not deliver proper permissions (`false`, even though it should be `true`). If the document gets _updates_ the mails will get send properly.

@afabiani I would really appreciate it if you could take some time to check this.

afabiani commented:
@gannebamm ok, I'll manage to test this on a live system with a valid SMTP configuration

afabiani commented:
@gannebamm it should be good now, can you test it again when you have the chance?

gannebamm commented:
Documents do work now on upload, but Layers fail with:
```
Object None needs to be persisted first
Traceback (most recent call last):
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py", line 538, in get_or_create
    return self.get(**kwargs), False
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py", line 406, in get
    raise self.model.DoesNotExist(
geonode.layers.models.Layer.DoesNotExist: Layer matching query does not exist.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/geonode/PycharmProjects/geonode_upstream/geonode/layers/views.py", line 257, in layer_upload
    saved_layer = file_upload(
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/geonode/PycharmProjects/geonode_upstream/geonode/layers/utils.py", line 692, in file_upload
    layer, created = Layer.objects.get_or_create(
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/modeltranslation/manager.py", line 370, in get_or_create
    return super(MultilingualQuerySet, self).get_or_create(**kwargs)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py", line 541, in get_or_create
    return self._create_object_from_params(kwargs, params)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py", line 575, in _create_object_from_params
    obj = self.create(**params)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/modeltranslation/manager.py", line 362, in create
    return super(MultilingualQuerySet, self).create(**kwargs)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/django/db/models/query.py", line 422, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/geonode/PycharmProjects/geonode_upstream/geonode/base/models.py", line 908, in save
    self.set_workflow_perms(published=True)
  File "/home/geonode/PycharmProjects/geonode_upstream/geonode/security/models.py", line 377, in set_workflow_perms
    assign_perm(perm,
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/guardian/shortcuts.py", line 135, in assign_perm
    return model.objects.assign_perm(perm, group, obj)
  File "/home/geonode/.virtualenvs/geonode_upstream/lib/python3.8/site-packages/guardian/managers.py", line 35, in assign_perm
    raise ObjectNotPersisted("Object %s needs to be persisted first"
guardian.exceptions.ObjectNotPersisted: Object None needs to be persisted first
```

This is way more complicated than I expected.

afabiani commented:
uhm, ok, do you have the workflow enabled? That could be a sort of unchecked exception side-effect.

I'll commit a patch asap.

afabiani commented:
@gannebamm I pushed a patch that should fix this last (hopefully) issue. Please check again. Thanks!

gannebamm commented:
> 
> uhm, ok, do you have the workflow enabled? That could be a sort of unchecked exception side-effect.

yes, I had that enabled. Seems like we caught a possible issue there. It is working now, but I need to test some more. I do not want to create any regressions here.
A question regarding comments:
`profiles = profiles.filter(Q(pk=instance_owner.pk) | Q(is_superuser=True))` - instance_owner is the comments owner. Therefore the account writing the comment will get notified. I do not think that is the way it should work. I would advise to use: `profiles = profiles.filter(Q(pk=resource.owner.pk) | Q(is_superuser=True))` instead, which will notify the owner of the layer/document a comment was written about.

Otherwise, the patch seems to work. But as said, please do not merge it, yet.

afabiani commented:
@gannebamm 

1. Sure, let's wait to merge until it is well tested. I would probably add more specific test cases in order to avoid regressions in the future.

2. The comment owner should not be notified, you are correct. Will look into that too.

gannebamm commented:
I have fixed the comment notifications and removed notifications for the resource owner. The owner must not be notified about his/her own layer being uploaded. Had forgotten to also except their own resource beeing updated. We can not check if the resource gets updates by themselves, but I think that is ok.

My tests run fine. I think this could be merged.

afabiani commented on geonode/notifications_helper.py:
```diff
@@ -129,7 +129,7 @@ def get_notification_recipients(notice_type_label, exclude_user=None, resource=N
                     exclude_users_ids.append(user.id)
                 if user.pk == resource.owner.pk and \
-                        not notice_type_label.endswith("_comment"):
+                        not (notice_type_label.endswith("_comment") or notice_type_label.endswith("_updated")):
```
Strange, that case should be already covered by `exclude_users_ids.append(exclude_user.id)`. Did you experienced such misbehavior?

gannebamm replied:
@afabiani Good point. It currently is not used in e.g. layer creation and will be `none` but I think it would be better to use that to exclude the signal firing user (= the one uploading, updateding, commenting).
